File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        res = new ArrayList<List<Integer>>();
        List<Integer> curr = new ArrayList<>();
        backTracking(nums, target, curr, 0);
        return res;
    }
    public void backTracking(int[] nums, int target, List<Integer> curr, int i){
        if(target == 0){
            res.add(new ArrayList(curr));
            return;
        }
        if(target < 0 || i >= nums.length) return;
        
        curr.add(nums[i]);
        backTracking(nums, target - nums[i], curr, i);
        curr.remove(curr.size() - 1);
        backTracking(nums, target, curr, i + 1);
    }
}

File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def binary_search(self, l: int, r: int, nums: List[int], target: int) -> int:
        if l > r:
            return -1
        middle = l + (r-l) // 2

        if nums[middle] == target:
            return middle
        if nums[middle] < target:
            return self.binary_search(middle+1, r, nums, target)
        
        return self.binary_search(l, middle-1, nums, target)
    def search(self, nums: List[int], target: int) -> int:
        return self.binary_search(0, len(nums)-1, nums, target)